python/taichi/lang/shell.py

```diff
@@ -0,0 +1,221 @@
+
+        if self.name is None:
+            # `inspect` for "Python script"
+            import inspect
+            self.getsource = inspect.getsource
+            self.getsourcelines = inspect.getsourcelines
+            self.getsourcefile = inspect.getsourcefile
+
+        elif self.name == ShellType.NATIVE:
+            # `dill.source` for "Python native shell"
+            import dill
+            self.getsource = dill.source.getsource
+            self.getsourcelines = dill.source.getsourcelines
+            self.getsourcefile = dill.source.getsourcefile
+
+        elif self.name == ShellType.IDLE:
+            # `.tmp_idle_source` for "Python IDLE shell"
+            # Thanks to IDLE's lack of support with `inspect`,
+            # we have to use a dirty hack to support Taichi there.
+            self.cache = {}
+
+            ppid_file = '.tmp_idle_ppid_' + str(os.getppid())
+            with open(ppid_file, 'w') as f:
+                print(f'[Taichi] touching {ppid_file}')
+                f.write('taichi')
+
+            def getsource(o):
+                func_name = o.__name__
+                if func_name in self.cache:
+                    return self.cache[func_name]
+                src = None
+                try:
+                    with open('.tmp_idle_source') as f:
+                        src = f.read()
+                except FileNotFoundError as e:
+                    _show_idle_error_message()
+                    raise e
+
+                # If user added our 'hacker-code' correctly,
+                # then the content of .tmp_idle_source should be:
+                #
+                # ===
+                # import taichi as ti
+                #
+                # ===
+                # @ti.kernel
+                # def func():    # x.find('def ') locate to here
+                #     pass
+                #
+                # ===
+                # func()
+                #
+                #
+                # Thanking IDLE dev :( It works anyway :)
+                for x in src.split('==='):
+                    x = x.strip()
+                    i = x.find('def ')
+                    if i == -1:
+                        continue
+                    name = x[i + 4:].split(':', maxsplit=1)[0]
+                    name = name.split('(', maxsplit=1)[0]
+                    if name.strip() == func_name:
+                        self.cache[func_name] = x
+                        return x
+                else:
+                    raise NameError(f'Could not find source for {o.__name__}!')
+
+            def getsourcelines(o):
+                lineno = 2  # TODO: consider include lineno in .tmp_idle_source?
+                lines = getsource(o).split('\n')
+                return lines, lineno
+
+            def getsourcefile(o):
+                return '<IDLE>'
+
+            self.getsource = getsource
+            self.getsourcelines = getsourcelines
+            self.getsourcefile = getsourcefile
```

Comment:
What's the best way to do this? `def self.getsourcelines()`? I only think about class member hooks, sorry about my poor python knowledge..